src/bootchooser.c

```diff
@@ -571,6 +617,45 @@ static gboolean uboot_set_state(RaucSlot *slot, gboolean good, GError **error)
 	return res;
 }
 
+/* Get slot marked as primary one */
+static RaucSlot* uboot_get_primary(GError **error)
+{
+	g_autoptr(GString) order = NULL;
+	g_auto(GStrv) bootnames = NULL;
+	GError *ierror = NULL;
+	RaucSlot *primary = NULL;
+	RaucSlot *slot;
+	GHashTableIter iter;
+	gboolean res;
+
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	res = uboot_env_get("BOOT_ORDER", &order, &ierror);
+	if (!res) {
+		g_propagate_error(error, ierror);
+		goto out;
+	}
+
+	/* Iterate over current boot order */
+	bootnames = g_strsplit(order->str, " ", -1);
+	for (gchar **bootname = bootnames; *bootname; bootname++) {
+		/* find matching slot entry */
+		g_hash_table_iter_init(&iter, r_context()->config->slots);
+		while (g_hash_table_iter_next(&iter, NULL, (gpointer*) &slot)) {
+			if (g_strcmp0(*bootname, slot->bootname) == 0) {
+				primary = slot;
```

Comment:
It won't be sufficient here to look if the key is in `BOOT_ORDER`. Consider the case we have `BOOT_ORDER=A B` and `BOOT_A_DEFAULT=0`, `BOOT_B_DEFAULT=3`. In this case, the boot selection won't boot A but B instead and thus we cannot return A as the 'primary' slot from this function, too.